registry/client/transport/http_reader.go

```diff
@@ -2,15 +2,31 @@ package transport
 
 import (
 	"bufio"
-	"bytes"
 	"errors"
 	"fmt"
 	"io"
-	"io/ioutil"
 	"net/http"
 	"os"
 )
 
+// HTTPError encapsulates an unsuccessful HTTP status code.
+type HTTPError struct {
+	Resp       *http.Response
+	bodyClosed bool
+}
+
+// Error renders an HTTPError to a string.
+func (h *HTTPError) Error() string {
+	// If the Error method gets called, the error propagated somewhere
+	// where it's not being handled natively as a HTTPError, so we need
+	// to take care of closing the body.
+	if !h.bodyClosed {
+		h.Resp.Body.Close()
```

Comment:
I don't like it much either.

The problem is that there may be many HTTP requests created by each `HTTPReadSeeker`. Each call to `Read` can potentially cause a new HTTP request, whose response body needs to be closed. That's why I was associating them with the error object.

How about keeping a slice in the `httpReadSeeker` struct that tracks every response body that was associated with an error condition, and closing all of those when the `Close` method is called?